### ISSUE TITLE
10th release goals

### ISSUE THREAD

**Issue by Actimel** [keboola/db-extractor-common#70]
Since we found some inconsistencies and blind spots i test I want to give it a deeper look when new major version will be released.

Already started with #68 (Updating dependencies), which possibly might contain some BC changes.

Then I'd like to add datadir tests to be more sure for future rework. Eventually also start with unit tests, but these might be added later (started a bit here #69).

When we have a good coverage by datadir tests would be nice to make a bit of refactoring. set stric types and cut many of code into smaller pieces. Then we could improve ConfigDefinitions checks (that caused a many troubles when we were setting up UI for terradata, especially when we broke a bit `db-ex-mysql` since it only one extractor using ConfigRows).

**Comment by pivnicek** [comment]
I thought type hints and return types are everywhere already
I'm not sure about removing `Pimple`, it's been good to us:)  [the if it ain't broke don't fix it rule], though I'd be happy to hear arguments for removing it.

What I'd like to see is:

- [ ] only use json (bye yaml for good)
- [ ] only support configRows

**Comment by odinuv** [comment]
just out of curiosity: what's pimple good for?

**Comment by pivnicek** [comment]
nostalgia, mostly.

**Comment by odinuv** [comment]
+ add a table object on init so as not to pass associative arrays (with deeply nested keys) and avoid https://github.com/keboola/db-extractor-mysql/issues/66 - inspiration https://github.com/keboola/docker-bundle/blob/master/src/Docker/Component.php ?

**Comment by Actimel** [comment]
Pimple has gone away but I can require it back if we really miss it although it did nothing. :D

**Comment by pivnicek** [comment]
I think it might be good to make a 10.0 branch with all these PRs combined when ready.  It'll be easier (at least for me) to review as a whole.
But in general I'm very happy to see it move in this direction, the datadir tests are really great.

**Comment by Actimel** [comment]
Yeah, that would be probably the best way. At first merge everything into one branch and then make a big (bang) merge.

**Comment by pivnicek** [comment]
So, from our discussion, it would probably be best to continue non-config row support in this version.  
But not yaml

**Comment by Actimel** [comment]
Yeah unfortunately... Other extractor except mysql are using ConfigRows too? Or there are some which doesn't support ConfigRows at all? Is there any way how to get how many projects are using old configuration?

**Comment by pivnicek** [comment]
Yes, currently only mysql and teradata support config rows.
The list is here. https://github.com/keboola/kbc-ui/blob/master/src/scripts/modules/ex-db-generic/actionsProvisioning.js#L55
When another component gets updated to support it, it can be added to this list and then the front end will support config rows for it, including migration.

**Comment by Actimel** [comment]
So that migration in mysql will be usable in other extractors too? Do we know how many mysql extractors is still using old config?

**Comment by pivnicek** [comment]
Yes, it's component agnostic.  It only cares that it has a job to convert `tables[]` into rows:)

I will make a transformation to monitor the #s of different style of config usages.  Will update here with a link when done

**Comment by Actimel** [comment]
That's nice. But still we need to support old config till all components are updated.

Now I'm thinking if I should revert that removing ConfigRows. Maybe creating some adapter from config to ConfigRows will be much better. We won't need separate validation but I'm not sure this might be possible.

**Comment by pivnicek** [comment]
hmmm, I don't think that the adapter will work because each run is only one job.  So if multiple tables[] come, there is no way to make multiple rows (ie >1 job)

**Comment by Actimel** [comment]
we could iterate those tables and for each table create temporary ConfigRows, then run it with it, if it is possible. I'll check it.

**Comment by pivnicek** [comment]
@Actimel I'm wondering if we can include solving this as a 10 goal https://github.com/keboola/db-extractor-common/issues/60

**Comment by odinuv** [comment]
would be great, because that is a big pain affecting all extractors
